aten/src/ATen/core/dispatch/OperatorEntry.cpp

```diff
@@ -153,51 +153,89 @@ const KernelFunction& OperatorEntry::computeDispatchTableEntry(const c10::Dispat
   //          if there's no direct registration to the backend key.
   //          Tensor factory functions used to have no registration to Autograd key but only to catchAll.
   //          In the past we directly call into backends(filled with catchAll) after BackendSelect.
   //          Now that we first call Autograd backend keys after BackendSelect, we should fill those
   //          with catchAll as well.
   //  (3) Use fallthrough kernel that are registered as fallback.
   //  (4) Use catchAll kernel if available
-  // TODO: currently Autograd is the only alias key, we'll update alias key precedence after we add new
-  //      alias keys AutogradDispatchCPUOrCUDA and Math.
-  // TODO: we can fix (2.2) and remove (4) after TypeDefault registrations are moved from catchAll to Math
+  // Alias Key Precedence:
+  //   Math > Autograd
+  // TODO: Update alias key precedence after we add new alias keys AutogradDispatchCPUOrCUDA .
+  // TODO: we can remove (2.3) and (4) after TypeDefault registrations are moved from catchAll to Math
   //       so that Math can populate to Autograd backend keys before fallback kernels.
 
   // 1. Operator registration
-  auto kern_it = kernels_.find(dispatch_key);
-  if (kern_it != kernels_.end()) {
-    TORCH_INTERNAL_ASSERT(!kern_it->second.empty());
-    TORCH_INTERNAL_ASSERT(kern_it->second.front().kernel.isValid());
-    return {kern_it->second.front(), "kernel"};
-
-  } else if (isIncludedInAlias(dispatch_key, DispatchKey::Autograd)) {
-    // 2.1. For autograd backend keys, use kernel from DispatchKey::Autograd if available
-    auto kern_autograd = kernels_.find(DispatchKey::Autograd);
-    if (kern_autograd != kernels_.end()) {
-      TORCH_INTERNAL_ASSERT(!kern_autograd->second.empty());
-      TORCH_INTERNAL_ASSERT(kern_autograd->second.front().kernel.isValid());
-      return {kern_autograd->second.front(), "autograd kernel"};
-
-    // 2.2. For autograd backend keys, we use kernel from alias Math key (catchAll will be moved to Math)
-    //      if there's no direct registration to the backend key.
-    } else if (kernels_.find(getBackendKeyFromAutograd(dispatch_key)) == kernels_.end()
-            && !catchAllKernel_.empty()) {
-      TORCH_INTERNAL_ASSERT(catchAllKernel_.front().kernel.isValid());
-      return {catchAllKernel_.front(), "catch all"};
+  auto direct_registration = getKernelForDispatchKey(dispatch_key);
+  if (direct_registration) {
+    return {*direct_registration.value(), "kernel"};
+  }
+
+  bool is_autograd_key_with_backend_kernel =
+    hasKernelForDispatchKeySet(getBackendKeySetFromAutograd(dispatch_key));
+  // 2.1. Use Math kernel if available. For autograd keys, we only use kernel from Math
+  //      when there's no direct registration to its corresponding backend key.
+  //      For AutogradOther, we return missing kernel if there's registration to any of
+  //      its backends.
+  auto math_registration = getKernelForDispatchKey(DispatchKey::Math);
```

Comment:
Small opt: do the `math_registration` lookup inside the if statement, as you can skip it entirely if the key is not in the alias. We should care a little about this as this code is somewhat hot in the static initialization phase.